### PR TITLE
[SPARK-19832][SQL]DynamicPartitionWriteTask get partitionPath should escape the partition name

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

Currently in DynamicPartitionWriteTask, when we get the paritionPath of a parition, we just escape the partition value, not escape the partition name.

this will cause some problems for some  special partition name situation, for example :
1) if the partition name contains '%' etc,  there will be two partition path created in the filesytem, one is for escaped path like '/path/a%25b=1', another is for unescaped path like '/path/a%b=1'.
and the data inserted stored in unescaped path, while the show partitions table will return 'a%25b=1' which the partition name is escaped. So here it is not consist. And I think the data should be stored in the escaped path in filesystem, which Hive2.0.0 also have the same action.

2) if the partition name contains ':', there will throw exception that new Path("/path","a:b"), this is illegal which has a colon in the relative path.

```
java.lang.IllegalArgumentException: java.net.URISyntaxException: Relative path in absolute URI: a:b
  at org.apache.hadoop.fs.Path.initialize(Path.java:205)
  at org.apache.hadoop.fs.Path.<init>(Path.java:171)
  at org.apache.hadoop.fs.Path.<init>(Path.java:88)
  ... 48 elided
Caused by: java.net.URISyntaxException: Relative path in absolute URI: a:b
  at java.net.URI.checkPath(URI.java:1823)
  at java.net.URI.<init>(URI.java:745)
  at org.apache.hadoop.fs.Path.initialize(Path.java:202)
  ... 50 more
```
## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by windpiger** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -1996,4 +1994,27 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
       }
     }
   }
+
+  Seq("a b", "a:b", "a%b", "a,b").foreach { specialCharInLoc =>
+    test(s"partition name of datasource table contains $specialCharInLoc") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: `data source table: partition column name containing $specialCharInLoc`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 2]
```diff
@@ -1996,4 +1994,27 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
       }
     }
   }
+
+  Seq("a b", "a:b", "a%b", "a,b").foreach { specialCharInLoc =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: just say `specialChars`

**Comment by cloud-fan** [comment]
good catch! LGTM

**Comment by cloud-fan** [comment]
now it conflicts...

**Comment by cloud-fan** [comment]
LGTM, pending tests

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**